chia/wallet/wallet_node.py

```diff
@@ -1022,25 +1008,29 @@ def get_full_node_peers_in_order(self) -> List[WSChiaConnection]:
                 neither.append(node)
         return synced_and_trusted + synced + trusted + neither
 
-    async def get_timestamp_for_height(self, height: uint32) -> uint64:
+    async def get_timestamp_for_height_from_peer(self, height: uint32, peer: WSChiaConnection) -> Optional[uint64]:
         """
         Returns the timestamp for transaction block at h=height, if not transaction block, backtracks until it finds
         a transaction block
         """
-        for cache in self.peer_caches.values():
-            cache_ts: Optional[uint64] = cache.get_height_timestamp(height)
-            if cache_ts is not None:
-                return cache_ts
+        cache = self.get_cache_for_peer(peer)
+        cached_timestamp = cache.get_height_timestamp(height)
+        if cached_timestamp is not None:
+            return cached_timestamp
 
-        for peer in self.get_full_node_peers_in_order():
-            last_tx_block = await fetch_last_tx_from_peer(height, peer)
-            if last_tx_block is None:
-                continue
+        last_tx_block = await fetch_last_tx_from_peer(height, peer)
+        if last_tx_block is None:
+            return None
 
-            assert last_tx_block.foliage_transaction_block is not None
-            self.get_cache_for_peer(peer).add_to_blocks(last_tx_block)
-            return last_tx_block.foliage_transaction_block.timestamp
+        assert last_tx_block.foliage_transaction_block is not None
+        self.get_cache_for_peer(peer).add_to_blocks(last_tx_block)
+        return last_tx_block.foliage_transaction_block.timestamp
 
+    async def get_timestamp_for_height(self, height: uint32) -> uint64:
+        for peer in self.get_full_node_peers_in_order():
+            timestamp = await self.get_timestamp_for_height_from_peer(height, peer)
```

Comment:
it's not obvious that picking the first hit is a good strategy. What if you have a trusted peer with this information, shouldn't you prefer that?

Reply:
The first might be a trusted one, see https://github.com/Chia-Network/chia-blockchain/blob/208485075c335cfe8a715fad10192b68a0afca25/chia/wallet/wallet_node.py#L988-L1013

Reply:
I see. that's quite subtle.